Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCell.cs

```diff
@@ -0,0 +1,1265 @@
+
+        /// <summary>
+        /// Gets the style applied to the cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> that includes the style settings of the cell inherited from the cell's parent row, column, and <see cref="T:System.Windows.Forms.DataGridView"/>.
+        /// </returns>
+        /// <param name="inheritedCellStyle">A <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> to be populated with the inherited cell style. </param><param name="rowIndex">The index of the cell's parent row. </param><param name="includeColors">true to include inherited colors in the returned cell style; otherwise, false. </param><exception cref="T:System.InvalidOperationException">The cell has no associated <see cref="T:System.Windows.Forms.DataGridView"/>.-or-<see cref="P:System.Windows.Forms.DataGridViewCell.ColumnIndex"/> is less than 0, indicating that the cell is a row header cell.</exception><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="rowIndex"/> is less than 0, or greater than or equal to the number of rows in the parent <see cref="T:System.Windows.Forms.DataGridView"/>.</exception>
+        public virtual DataGridViewCellStyle GetInheritedStyle(DataGridViewCellStyle inheritedCellStyle, int rowIndex,
+            bool includeColors)
+        {
+            Contract.Requires(this.DataGridView != null);
+            Contract.Requires(rowIndex >= 0 || rowIndex < this.DataGridView.Rows.Count);
+            Contract.Requires(this.ColumnIndex < 0);
+            return default(DataGridViewCellStyle);
+        }
+
+        /// <summary>
+        /// Calculates the preferred size, in pixels, of the cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Drawing.Size"/> that represents the preferred size, in pixels, of the cell.
+        /// </returns>
+        /// <param name="graphics">The <see cref="T:System.Drawing.Graphics"/> used to draw the cell.</param><param name="cellStyle">A <see cref="T:System.Windows.Forms.DataGridViewCellStyle"/> that represents the style of the cell.</param><param name="rowIndex">The zero-based row index of the cell.</param><param name="constraintSize">The cell's maximum allowable size.</param>
+        // protected virtual Size GetPreferredSize(Graphics graphics, DataGridViewCellStyle cellStyle, int rowIndex, Size constraintSize)
+        
+        /// <summary>
+        /// Gets the size of the cell.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// A <see cref="T:System.Drawing.Size"/> representing the cell's dimensions.
+        /// </returns>
+        /// <param name="rowIndex">The index of the cell's parent row.</param><exception cref="T:System.InvalidOperationException"><paramref name="rowIndex"/> is -1</exception>
+        protected virtual Size GetSize(int rowIndex)
+        {
+            Contract.Requires(rowIndex != -1);
```

Comment:
This precondition is also not correct. See http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/DataGridViewCell.cs,71451549454cc494

Should be:

`Contract.Requires(DataGridView == null || rowIndex != -1);`